File: app/src/main/java/com/omar/retromp3recorder/app/di/VoiceRecorder.java
package com.omar.retromp3recorder.app.di;

import io.reactivex.Observable;

public interface VoiceRecorder {

    Observable<VoiceRecorder.Event> observeEvents();

    void record(RecorderProps props);

    void stopRecord();

    enum SampleRate {
        _44100(44100),
        _22050(22050),
        _11052(11052),
        _8000(8800);

        public final int value;

        SampleRate(int value) {
            this.value = value;
        }
    }

    enum BitRate {
        _320(320), _192(192), _160(160), _128(128);
        public final int value;

        BitRate(int value) {
            this.value = value;
        }
    }

    interface Event {
    }

    final class Message implements Event {
        public final String message;

        public Message(String message) {
            this.message = message;
        }
    }

    final class Error implements Event {
        public final String error;

        public Error(String error) {
            this.error = error;
        }
    }

    final class RecorderProps {
        public final String filepath;
        public final BitRate bitRate;
        public final SampleRate sampleRate;

        public RecorderProps(String filepath, BitRate bitRate, SampleRate sampleRate) {
            this.filepath = filepath;
            this.bitRate = bitRate;
            this.sampleRate = sampleRate;
        }
    }
}

File: app/src/main/java/com/omar/retromp3recorder/app/recording/recorder/VoiceRecorder.kt
package com.omar.retromp3recorder.app.recording.recorder

import android.media.AudioFormat
import com.github.alkurop.stringerbell.Stringer
import io.reactivex.Observable

interface VoiceRecorder {
    fun observeEvents(): Observable<Event>
    fun record(props: RecorderProps)
    fun stopRecord()
    fun isRecording(): Boolean

    enum class SampleRate(val value: Int) {
        _44100(44100), _22050(22050), _11025(11025), _8000(8000);
    }

    enum class BitRate(val value: Int) {
        _320(320), _192(192), _160(160), _128(128);
    }

    sealed class Event {
        data class Message(val message: Stringer) : Event()
        data class Error(val error: Stringer) : Event()
    }

    data class RecorderProps(
        val filepath: String,
        val bitRate: BitRate,
        val sampleRate: SampleRate
    )

    companion object {
        val AUDIO_FORMAT_PRESETS = shortArrayOf(
            AudioFormat.ENCODING_PCM_8BIT.toShort(),
            AudioFormat.ENCODING_PCM_16BIT
                .toShort()
        )
        val QUALITY_PRESETS = intArrayOf(
            2,
            5,
            7
        ) // the lower the better
        val CHANNEL_PRESETS = shortArrayOf(
            AudioFormat.CHANNEL_IN_MONO.toShort(),
            AudioFormat.CHANNEL_IN_STEREO
                .toShort()
        )
    }
}
